gcloud-java-dns/src/main/java/com/google/gcloud/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1426 @@
+      this.code = code;
+      this.domain = domain;
+      this.reason = reason;
+      this.status = status;
+    }
+
+    Response response(String message) {
+      try {
+        return new Response(code, toJson(message));
+      } catch (IOException e) {
+        return Error.INTERNAL_ERROR.response("Error when generating JSON error response.");
+      }
+    }
+
+    private String toJson(String message) throws IOException {
+      Map<String, Object> errors = new HashMap<>();
+      errors.put("domain", domain);
+      errors.put("message", message);
+      errors.put("reason", reason);
+      Map<String, Object> args = new HashMap<>();
+      args.put("errors", ImmutableList.of(errors));
+      args.put("code", code);
+      args.put("message", message);
+      args.put("status", status);
+      return jsonFactory.toString(ImmutableMap.of("error", args));
+    }
+  }
+
+  private class RequestHandler implements HttpHandler {
+
+    /**
+     * Chooses the proper handler for a request.
+     */
+    private Response pickHandler(HttpExchange exchange, CallRegex regex) {
+      switch (regex) {
+        case CHANGE_GET:
+          return handleChangeGet(exchange);
+        case CHANGE_LIST:
+          return handleChangeList(exchange);
+        case ZONE_GET:
+          return handleZoneGet(exchange);
+        case ZONE_DELETE:
+          return handleZoneDelete(exchange);
+        case ZONE_LIST:
+          return handleZoneList(exchange);
+        case PROJECT_GET:
+          return handleProjectGet(exchange);
+        case RECORD_LIST:
+          return handleDnsRecordList(exchange);
+        case ZONE_CREATE:
+          try {
+            return handleZoneCreate(exchange);
+          } catch (IOException ex) {
+            return Error.BAD_REQUEST.response(ex.getMessage());
+          }
+        case CHANGE_CREATE:
+          try {
+            return handleChangeCreate(exchange);
+          } catch (IOException ex) {
+            return Error.BAD_REQUEST.response(ex.getMessage());
+          }
+        default:
+          return Error.INTERNAL_ERROR.response("Operation without a handler.");
+      }
+    }
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
+      String requestMethod = exchange.getRequestMethod();
+      String rawPath = exchange.getRequestURI().getRawPath();
+      for (CallRegex regex : CallRegex.values()) {
+        if (requestMethod.equals(regex.method) && rawPath.matches(regex.pathRegex)) {
+          // there is a match, pass the handling accordingly
+          Response response = pickHandler(exchange, regex);
+          writeResponse(exchange, response);
+          return; // only one match is possible
+        }
+      }
+      // could not be matched, the service returns 404 page not found here
+      writeResponse(exchange, Error.NOT_FOUND.response("The url does not match any API call."));
+    }
+
+    private Response handleZoneDelete(HttpExchange exchange) {
```

Comment:
Since the body of this method and the other handlers below have a lot of common code, I think we could do some refactoring to share path relativization, project ID extraction, and optional zone/query parameter extraction.